### PR TITLE
IGNITE-12108 TCP Communication Metrics ported to a new framework.

### PR DESCRIPTION
**Pull request by ibessonov**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridAbstractCommunicationClient.java [hunk 1]
```diff
@@ -32,19 +39,27 @@
...
+    /** Sent bytes count metric. */
+    @Nullable protected final LongAdderMetric sentBytesCntMetric;
+
     /**
      * @param connIdx Connection index.
-     * @param metricsLsnr Metrics listener.
+     * @param mreg Metrics registry.
      */
-    protected GridAbstractCommunicationClient(int connIdx, @Nullable GridNioMetricsListener metricsLsnr) {
+    protected GridAbstractCommunicationClient(int connIdx, @Nullable MetricRegistry mreg) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridAbstractCommunicationClient.java]: Why `mreg` can be null?

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java [hunk 2]
```diff
@@ -293,12 +320,12 @@ private GridNioServer(
         int sndQueueLimit,
         boolean directMode,
         boolean daemon,
-        GridNioMetricsListener metricsLsnr,
         GridNioMessageWriterFactory writerFactory,
         IgnitePredicate<Message> skipRecoveryPred,
         IgniteBiInClosure<GridNioSession, Integer> msgQueueLsnr,
         boolean readWriteSelectorsAssign,
         @Nullable GridWorkerListener workerLsnr,
+        @Nullable MetricRegistry mreg,
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java]: Why `mreg` can be null?

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java [hunk 3]
```diff
@@ -2860,14 +2895,17 @@ final void reset0() {
...
     public int outboundMessagesQueueSize() {
-        int res = 0;
-
-        for (GridSelectorNioSessionImpl ses : sessions)
-            res += ses.writeQueueSize();
+        if (mreg == null)
+            return -1;
 
-        return res;
+        return (int) mreg.longAdderMetric(
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java]: We should store `LongMetric` inside object instance.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 4]
```diff
@@ -17,65 +17,125 @@
...
+        sentMsgsCntByTypeMetricFactory = directType -> mreg.longAdderMetric(
+            sentMessagesByTypeMetricName(directType),
+            SENT_MESSAGES_BY_TYPE_METRIC_DESC
+        );
+        rcvdMsgsCntByTypeMetricFactory = directType -> mreg.longAdderMetric(
+            receivedMessagesByTypeMetricName(directType),
+            RECEIVED_MESSAGES_BY_TYPE_METRIC_DESC
+        );
+
+        sentMsgsCntByNodeIdMetricFactory = nodeId -> mmgr.registry(COMMUNICATION_METRICS_GROUP_NAME + "." + nodeId)
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: Let's use `MetricUtils#metricName` instead of `COMMUNICATION_METRICS_GROUP_NAME + "." + nodeId`

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java [hunk 5]
```diff
@@ -403,6 +401,50 @@
     /** Handshake wait message type. */
     public static final short HANDSHAKE_WAIT_MSG_TYPE = -28;
 
+    /** Communication metrics group name. */
+    public static final String COMMUNICATION_METRICS_GROUP_NAME = "communication.tcp";
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java]: Let's use `MetricUtils#metricName` instead of `"communication.tcp"`.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java [hunk 6]
```diff
@@ -403,6 +401,50 @@
...
+
+    /** */
+    public static final String RECEIVED_MESSAGES_METRIC_NAME = "receivedMessagesCount";
+
+    /** */
+    public static final String RECEIVED_MESSAGES_METRIC_DESC = "Total number of messages received by current node";
+
+
+    /** */
+    public static String sentMessagesByTypeMetricName(Short directType) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java]: Let's declare all constants, first, and methods after.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java [hunk 7]
```diff
@@ -403,6 +401,50 @@
...
+        return "sentMessagesByType." + directType;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java]: Let's use MetricUtils#metricName instead of this and similar methods.

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java [hunk 8]
```diff
@@ -93,19 +97,22 @@ public IgniteTestResources() throws IgniteCheckedException {
...
 
     /**
      * @param cfg Ignite configuration
      */
     public IgniteTestResources(IgniteConfiguration cfg) throws IgniteCheckedException {
         this.cfg = cfg;
-        this.log = rootLog.getLogger(getClass());
-        this.jmx = prepareMBeanServer();
-        this.rsrcProc = new GridResourceProcessor(new GridTestKernalContext(this.log, this.cfg));
+        log = rootLog.getLogger(getClass());
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java]: why this change?

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java [hunk 9]
```diff
@@ -115,8 +122,9 @@ public IgniteTestResources(MBeanServer jmx) throws IgniteCheckedException {
         assert jmx != null;
 
         this.jmx = jmx;
-        this.log = rootLog.getLogger(getClass());
-        this.rsrcProc = new GridResourceProcessor(new GridTestKernalContext(this.log));
+        log = rootLog.getLogger(getClass());
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java]: why this change?

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java [hunk 10]
```diff
@@ -126,8 +134,9 @@ public IgniteTestResources(IgniteLogger log) throws IgniteCheckedException {
         assert log != null;
 
         this.log = log.getLogger(getClass());
-        this.jmx = prepareMBeanServer();
-        this.rsrcProc = new GridResourceProcessor(new GridTestKernalContext(this.log));
+        jmx = prepareMBeanServer();
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java]: why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridAbstractCommunicationClient.java (hunk 1 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridAbstractCommunicationClient.java]: This class might be used by components other than TCP SPI. You can see that "GridNioMetricsListener" was nullable as well.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java (hunk 2 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java]: Same as above.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java (hunk 4 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: ok

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java (hunk 5 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java]: ok

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java (hunk 6 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java]: ok

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java (hunk 7 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationSpi.java]: ok

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java (hunk 8 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java]: "this." was not necessary so I removed it while adding new field.

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java (hunk 9 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java]: same as above.

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java (hunk 10 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java]: same

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java (hunk 3 above)
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java]: It's about unnecessary `Map#get` operation.
Every 'mreg.longAdderMetric` translates to `Map#get`.
Why do we need it? We can store the variable, isn't it?

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java (hunk 9 above)
Comment (nizhikov) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java]: Please, revert unnecessary changes.

File: modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java (hunk 10 above)
Comment (nizhikov) [reply modules/core/src/test/java/org/apache/ignite/testframework/junits/IgniteTestResources.java]: Please, revert unnecessary changes.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java (hunk 2 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java]: Let me please argue with you about this particular constructor =)
I totally forgot that it's a private constructor and only "GridNioServer.Builder#build" invokes it. There's no point in splitting it because no one will invoke other constructor with fewer parameters. Let's leave it as is.

Just in case - GridAbstractCommunicationClient constructor will be fixed as you asked. I't not about arguing for the sake of arguing.

Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java]: Thanks, Ilya. 
Please, let me know when PR becomes ready to be reviewed.

File: modules/core/src/test/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationStatisticsTest.java [hunk 11]
```diff
@@ -53,7 +60,7 @@
     private final Object mux = new Object();
 
     /** */
-    private final CountDownLatch latch = new CountDownLatch(1);
+    private CountDownLatch latch = new CountDownLatch(1);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationStatisticsTest.java]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 12]
```diff
@@ -17,74 +17,162 @@
...
+
+    /** Sent messages count metrics grouped by message type. */
+    ConcurrentHashMap<Short, AtomicLongMetric> sentMsgsMetricsByType = new ConcurrentHashMap<>();
+
+    /** Received messages count metrics grouped by message type. */
+    ConcurrentHashMap<Short, AtomicLongMetric> rcvdMsgsMetricsByType = new ConcurrentHashMap<>();
+
+    /**
+     * Sent messages count metrics grouped by message node id.
+     * @deprecated for JMX support only.
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: Let's start sentence with the capital letter and explain to the reader, what he(or she) should use instead of this map.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 13]
```diff
@@ -17,74 +17,162 @@
...
+     */
+    @Deprecated
+    ConcurrentHashMap<UUID, AtomicLong> sentMsgsMetricsByNodeId = new ConcurrentHashMap<>();
+
+    /**
+     * Received messages metrics count grouped by message node id.
+     * @deprecated for JMX support only.
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: Let's start sentence with the capital letter and explain to the reader, what he(or she) should use instead of this map.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 14]
```diff
@@ -17,74 +17,162 @@
...
 
-            allMetrics.add(metrics);
+    /** */
+    private final Function<Object, AtomicLongMetric> rcvdMsgsCntByConsistentIdMetricFactory;
 
-            return metrics;
-        }
-    };
+    /** Sent bytes count metric.*/
+    private final AtomicLongMetric sentBytesMetric;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: Let's keep `LongAdder` as an internal storage for all metrics that was based on it previously.
You can use `MetricRegistry#longAdderMetric` for it.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 15]
```diff
@@ -17,74 +17,162 @@
...
+    private final GridMetricManager mmgr;
 
-    /** Received bytes count. */
-    private final LongAdder rcvdBytesCnt = new LongAdder();
+    /** Metrics registry. */
+    private final MetricRegistry mreg;
 
-    /** Sent bytes count.*/
-    private final LongAdder sentBytesCnt = new LongAdder();
+    /** */
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: Please, add javadoc for this variable.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 16]
```diff
@@ -17,74 +17,162 @@
...
+    private final Function<Short, AtomicLongMetric> sentMsgsCntByTypeMetricFactory;
 
-    /** All registered metrics. */
-    private final Set<ThreadMetrics> allMetrics = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    /** */
+    private final Function<Short, AtomicLongMetric> rcvdMsgsCntByTypeMetricFactory;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: Please, add javadoc for this variable.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 17]
```diff
@@ -17,74 +17,162 @@
...
 
-    /** Thread-local metrics. */
-    private final ThreadLocal<ThreadMetrics> threadMetrics = new ThreadLocal<ThreadMetrics>() {
-        @Override protected ThreadMetrics initialValue() {
-            ThreadMetrics metrics = new ThreadMetrics();
+    /** */
+    private final Function<Object, AtomicLongMetric> sentMsgsCntByConsistentIdMetricFactory;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: Please, add javadoc for this variable.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 18]
```diff
@@ -17,74 +17,162 @@
...
+
+    /**
+     * Received messages metrics count grouped by message node id.
+     * @deprecated for JMX support only.
+     */
+    @Deprecated
+    ConcurrentHashMap<UUID, AtomicLong> rcvdMsgsMetricsByNodeId = new ConcurrentHashMap<>();
+
+    /** Sent messages count metrics grouped by message node consistent id. */
+    ConcurrentHashMap<Object, AtomicLongMetric> sentMsgsMetricsByConsistentId = new ConcurrentHashMap<>();
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: I'm not sure we should change the way we collect metrics in this PR.
Let's keep existing behaviour?

This also will simplify PR a lot.

**Comment by nizhikov** [comment]
@ibessonov Thanks for the PR update.

Looks much better now!
Please, take a look at my comments

File: modules/core/src/main/java/org/apache/ignite/internal/processors/resource/GridResourceProcessor.java [hunk 19]
```diff
@@ -74,6 +74,8 @@ public GridResourceProcessor(GridKernalContext ctx) {
             new GridResourceLoggerInjector(ctx.config().getGridLogger());
         injectorByAnnotation[GridResourceIoc.ResourceAnnotation.IGNITE_INSTANCE.ordinal()] =
             new GridResourceBasicInjector<>(ctx.grid());
+        injectorByAnnotation[GridResourceIoc.ResourceAnnotation.METRIC_MANAGER.ordinal()] =
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/resource/GridResourceProcessor.java]: Because supplier injector is lazy. In this particular place metric manager might not be initialized yet.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java (hunk 18 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: I changed that because in old implementation we stored basically the same map in each thread separately. That's the obvious answer.

The other thing is that old implementation had concurrency issues that no one cared about. Like broken values visibility щк potential ConcurrentModificationExceptions. I wanted new implementation to be correct.

File: modules/core/src/test/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationStatisticsTest.java (hunk 11 above)
Comment (ibessonov) [reply modules/core/src/test/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationStatisticsTest.java]: Artifact from test that didn't make it to the final version. I'll revert it.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridShmemCommunicationClient.java [hunk 20]
```diff
@@ -77,6 +86,9 @@ public GridShmemCommunicationClient(
         writeBuf.order(ByteOrder.nativeOrder());
 
         this.formatter = formatter;
+
+        sentBytesCntMetric = mreg == null ?
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridShmemCommunicationClient.java]: Good catch, thank you!

File: modules/core/src/main/java/org/apache/ignite/spi/communication/CommunicationSpi.java [hunk 21]
```diff
@@ -67,40 +67,53 @@
      * Gets sent messages count.
      *
      * @return Sent messages count.
+     * @deprecated Will be removed in the next major release and replaced with new metrics API.
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/CommunicationSpi.java]: I need you advice about the comment. Are there any plans to cancel JMX support for these values? I didn't think about it long enough. Maybe just remove that Deprecated annotation.
Anyway, specifying some metrics names here would be wrong, CommunicationSpi is interface after all and some implementations may want to use some other specific metric names, or no metrics at all.

File: modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridShmemCommunicationClient.java [hunk 22]
```diff
@@ -26,13 +26,19 @@
...
 import org.apache.ignite.internal.util.ipc.shmem.IpcSharedMemoryClientEndpoint;
 import org.apache.ignite.internal.util.lang.IgniteInClosure2X;
 import org.apache.ignite.internal.util.typedef.internal.S;
 import org.apache.ignite.internal.util.typedef.internal.U;
 import org.apache.ignite.lang.IgniteInClosure;
 import org.apache.ignite.plugin.extensions.communication.Message;
 import org.apache.ignite.plugin.extensions.communication.MessageFormatter;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.util.nio.GridNioServer.SENT_BYTES_METRIC_DESC;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridShmemCommunicationClient.java]: nit: unused import.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/CommunicationSpi.java [hunk 23]
```diff
@@ -100,6 +100,7 @@
 
     /**
      * Resets metrics for this SPI instance.
+     *
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/CommunicationSpi.java]: nit: unnecessary change.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java (hunk 18 above)
Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: You introduced metrics based on node consistency id.
I think we should avoid introducing new metrics here.

The only goal of this ticket is to migrate *existing* metrics to the new framework.

Let's 
1. remove metrics based on consistency id.
2. revert changes in `ConnectionKey` and `TcpCommunicationSpi` that made to support it.

**Comment by nizhikov** [comment]
@ibessonov 

1. In `TcpCommunicationMetricsListener` let's inline the following variables. They used only once in the code:

```
    /** Function to be used in {@link Map#computeIfAbsent(Object, Function)} of {@link #sentMsgsMetricsByType}. */
    private final Function<Short, LongAdderMetric> sentMsgsCntByTypeMetricFactory;

    /** Function to be used in {@link Map#computeIfAbsent(Object, Function)} of {@link #rcvdMsgsMetricsByType}. */
    private final Function<Short, LongAdderMetric> rcvdMsgsCntByTypeMetricFactory;

    /** Function to be used in {@link Map#computeIfAbsent(Object, Function)} of {@link #sentMsgsMetricsByNodeId}. */
    private final Function<UUID, LongAdderMetric> sentMsgsCntByNodeIdMetricFactory;

    /** Function to be used in {@link Map#computeIfAbsent(Object, Function)} of {@link #rcvdMsgsMetricsByNodeId}. */
    private final Function<UUID, LongAdderMetric> rcvdMsgsCntByNodeIdMetricFactory;
```

**Comment by nizhikov** [comment]
@ibessonov As we discussed privately we should resolve potential contention on `ConcurrentHashMap` access.

```
            sentMsgsMetricsByType.computeIfAbsent(msg.directType(), sentMsgsCntByTypeMetricFactory).increment();

            sentMsgsMetricsByNodeId.computeIfAbsent(nodeId, sentMsgsCntByNodeIdMetricFactory).increment();
```

We also should perform a yardstick check of these changes.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 24]
```diff
@@ -22,60 +22,145 @@
...
+
+        sentMsgsMetric = mreg.longAdderMetric(SENT_MESSAGES_METRIC_NAME, SENT_MESSAGES_METRIC_DESC);
+        rcvdMsgsMetric = mreg.longAdderMetric(RECEIVED_MESSAGES_METRIC_NAME, RECEIVED_MESSAGES_METRIC_DESC);
+    }
+
+    /** */
+    private static synchronized MetricRegistry getOrCreateMetricRegistry(GridMetricManager mmgr, UUID nodeId) {
+        String regName = MetricUtils.metricName(COMMUNICATION_METRICS_GROUP_NAME, nodeId.toString());
+
+        for (MetricRegistry mreg : mmgr) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: We can eliminate iteration on all registries if we use `GridMetricManager#addMetricRegistryCreationListener` to implement registry initialization logic.

```
        cctx.kernalContext().metric().addMetricRegistryCreationListener(mreg -> {
            if (!mreg.name().startsWith(COMMUNICATION_METRICS_GROUP_NAME))
                return;

            mreg.longAdderMetric(SENT_MESSAGES_BY_NODE_ID_METRIC_NAME, SENT_MESSAGES_BY_NODE_ID_METRIC_DESC);

            mreg.longAdderMetric(RECEIVED_MESSAGES_BY_NODE_ID_METRIC_NAME, RECEIVED_MESSAGES_BY_NODE_ID_METRIC_DESC);
        });
```

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 25]
```diff
@@ -247,10 +278,55 @@ public long receivedBytesCount() {
...
+                    Map<Short, String> msgTypMap0 = msgTypMap;
+
+                    if (msgTypMap0 != null) {
+                        String typeName = msgTypMap0.get(directType);
+
+                        if (typeName != null)
+                            res.put(typeName, ((LongMetric)metric).value());
+                    }
+                }
+                catch (NumberFormatException ignore) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: Let's print the warning to log here.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 26]
```diff
@@ -247,10 +278,55 @@ public long receivedBytesCount() {
...
+    }
+
+    /** */
+    protected Map<UUID, Long> collectMessagesCountByNodeId(String metricName) {
         Map<UUID, Long> res = new HashMap<>();
 
-        for (ThreadMetrics metrics : allMetrics)
-            addMetrics(res, metrics.sentMsgsCntByNode);
+        String mregPrefix = COMMUNICATION_METRICS_GROUP_NAME + MetricUtils.SEPARATOR;
+        for (MetricRegistry mreg : mmgr) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: empty line required

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java [hunk 27]
```diff
@@ -247,10 +278,55 @@ public long receivedBytesCount() {
...
+            if (mreg.name().startsWith(mregPrefix)) {
+                String nodeIdStr = mreg.name().substring(mregPrefix.length());
+
+                try {
+                    UUID nodeId = UUID.fromString(nodeIdStr);
+
+                    res.put(nodeId, mreg.<LongMetric>findMetric(metricName).value());
+                }
+                catch (IllegalArgumentException ignore) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: We shouldn't swallow exception here. Let's remove empty catch block.

File: modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java (hunk 25 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java]: I'll just remove it, it's excessive

**Comment by nizhikov** [comment]
LGTM.

@ibessonov  Thank you, so much for the valuable contribution!
I will gather performance tests results and merge this patch if everything is OK.
